### PR TITLE
SharedBuffer should be copied and then attributed back to WebContent

### PR DESCRIPTION
**Pull request by jyavenard**
#### fa2442c60ff100dde5874de0be9d57f59860000f
<pre>
SharedBuffer should be copied and then attributed back to WebContent
<a href="https://bugs.webkit.org/show_bug.cgi?id=240223">https://bugs.webkit.org/show_bug.cgi?id=240223</a>
rdar://91006582

Patch by Jean-Yves Avenard &lt;jean-yves.avenard@apple.com &gt; on 2022-05-14
Reviewed by Jer Noble.

Ideally, when mapping an IPCHandle we would prefer to use Copy On Write.
However, there&apos;s no support for COW with explicitly-owned memory such as
those created with VM_FLAGS_PURGABLE flag.
As such, when calling mach_vm_map with copy=true argument is set the
entire page will be copied.
It is therefore more performant to manually create a new buffer and copy
the actual content.
The memory can then be blamed on a particular process.

Rework FragmentedSharedBuffer/SharedBuffer IPC&apos;s serializer to perform
a copy, partially reverting changes made in bug 233401.

We replace the unnecessary use of SharedMemory::IPCHandle and directly
send SharedBuffer in clipboard and PDF related methods.
Rename SharedBufferCopy to SharedBufferReference to more accurately
describe what it does.

Make new use of the SharedBufferReference class, replacing the use of IPCHandle
in RemoteSourceBufferProxy and RemoteMediaResourceManager.

Rename SharedBufferReference::buffer() into
SharedBufferReference::unsafeBuffer() to give better clarity that accessing
SharedBuffer and SharedMemory should be treated with care.

SharedBuffer sent from trusted processes such as network process should
prefer the use of SharedBufferReference over directly using IPCHandle or
SharedBuffer, it makes optimisation easier and avoid duplicating code.

No observable change, covered by existing tests.

* GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::remoteMediaResourceManager):
* GPUProcess/media/RemoteCDMInstanceProxy.h:
* GPUProcess/media/RemoteImageDecoderAVFProxy.cpp:
(WebKit::RemoteImageDecoderAVFProxy::createDecoder):
(WebKit::RemoteImageDecoderAVFProxy::setData):
* GPUProcess/media/RemoteImageDecoderAVFProxy.h:
* GPUProcess/media/RemoteImageDecoderAVFProxy.messages.in:
* GPUProcess/media/RemoteMediaResourceManager.cpp:
(WebKit::RemoteMediaResourceManager::RemoteMediaResourceManager):
(WebKit::RemoteMediaResourceManager::dataReceived):
* GPUProcess/media/RemoteMediaResourceManager.h:
* GPUProcess/media/RemoteMediaResourceManager.messages.in:
* GPUProcess/media/RemoteSourceBufferProxy.cpp:
(WebKit::RemoteSourceBufferProxy::append):
* GPUProcess/media/RemoteSourceBufferProxy.h:
* GPUProcess/media/RemoteSourceBufferProxy.messages.in:
* NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::bufferingTimerFired):
(WebKit::NetworkResourceLoader::sendBuffer):
(WebKit::NetworkResourceLoader::dataReceivedThroughContentFilter):
* NetworkProcess/ServiceWorker/ServiceWorkerDownloadTask.cpp:
(WebKit::ServiceWorkerDownloadTask::didReceiveData):
* NetworkProcess/ServiceWorker/ServiceWorkerDownloadTask.h:
* NetworkProcess/ServiceWorker/ServiceWorkerDownloadTask.messages.in:
* NetworkProcess/ServiceWorker/ServiceWorkerFetchTask.cpp:
(WebKit::ServiceWorkerFetchTask::didReceiveData):
(WebKit::ServiceWorkerFetchTask::loadBodyFromPreloader):
* NetworkProcess/ServiceWorker/ServiceWorkerFetchTask.h:
* NetworkProcess/ServiceWorker/ServiceWorkerFetchTask.messages.in:
* Platform/IPC/SharedBufferCopy.cpp: Removed.
* Platform/IPC/SharedBufferCopy.h: Removed.
* Platform/IPC/SharedBufferReference.cpp: Added.
(IPC::SharedBufferReference::encode const):
(IPC::SharedBufferReference::decode):
(IPC::SharedBufferReference::unsafeBuffer const):
(IPC::SharedBufferReference::data const):
(IPC::SharedBufferReference::bufferWithOwner const):
* Platform/IPC/SharedBufferReference.h: Added.
(IPC::SharedBufferReference::SharedBufferReference):
(IPC::SharedBufferReference::size const):
(IPC::SharedBufferReference::isEmpty const):
(IPC::SharedBufferReference::isNull const):
* Platform/SharedMemory.cpp:
(WebKit::transferOwnershipOfMemory const):
* Platform/SharedMemory.h:
* Platform/cocoa/SharedMemoryCocoa.cpp:
(WebKit::changeOwnershipOfMemory):
(WebKit::SharedMemory::Handle::takeOwnershipOfMemory const):
(WebKit::SharedMemory::Handle::transferOwnershipOfMemory const):
(WebKit::machProtection):
(WebKit::SharedMemory::map):
* Scripts/process-entitlements.sh:
* Shared/RemoteLayerTree/CGDisplayListImageBufferBackend.cpp:
(WebKit::CGDisplayListImageBufferBackend::createBackendHandle const):
* Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::applyBackingStoreToLayer):
* Shared/WebCoreArgumentCoders.cpp:
(IPC::encodeSharedBuffer):
(IPC::decodeSharedBuffer):
* Sources.txt:
* UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView createPDFWithConfiguration:completionHandler:]):
* UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::platformRegisterAttachment):
* UIProcess/Cocoa/WebPasteboardProxyCocoa.mm:
(WebKit::WebPasteboardProxy::getPasteboardBufferForType):
(WebKit::WebPasteboardProxy::setPasteboardBufferForType):
(WebKit::WebPasteboardProxy::readBufferFromPasteboard):
* UIProcess/Cocoa/WebProcessProxyCocoa.mm:
(WebKit::WebProcessProxy::sendAudioComponentRegistrations):
* UIProcess/WebPageProxy.cpp:
(WebKit::CompletionHandler&lt;void):
(WebKit::WebPageProxy::drawPagesToPDF):
(WebKit::WebPageProxy::drawToPDF):
(WebKit::WebPageProxy::getLoadDecisionForIcon):
(WebKit::WebPageProxy::updateAttachmentAttributes):
(WebKit::WebPageProxy::registerAttachmentIdentifierFromData):
(WebKit::WebPageProxy::platformRegisterAttachment):
* UIProcess/WebPageProxy.h:
* UIProcess/WebPageProxy.messages.in:
* UIProcess/WebPasteboardProxy.cpp:
(WebKit::WebPasteboardProxy::readBufferFromPasteboard):
* UIProcess/WebPasteboardProxy.h:
* UIProcess/WebPasteboardProxy.messages.in:
* UIProcess/WebURLSchemeTask.cpp:
(WebKit::WebURLSchemeTask::didReceiveData):
* UIProcess/gtk/WebPasteboardProxyGtk.cpp:
(WebKit::WebPasteboardProxy::readBuffer):
* UIProcess/ios/WKContentView.mm:
(-[WKContentView _wk_pageCountForPrintFormatter:]):
* UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::drawToPDFiOS):
* UIProcess/mac/WebPageProxyMac.mm:
(WebKit::WebPageProxy::dataSelectionForPasteboard):
* WebKit.xcodeproj/project.pbxproj:
* WebProcess/GPU/graphics/ImageBufferBackendHandle.h:
* WebProcess/GPU/media/RemoteImageDecoderAVF.cpp:
(WebKit::RemoteImageDecoderAVF::setData):
* WebProcess/GPU/media/RemoteImageDecoderAVFManager.cpp:
(WebKit::RemoteImageDecoderAVFManager::createImageDecoder):
* WebProcess/GPU/media/RemoteMediaResourceProxy.cpp:
(WebKit::RemoteMediaResourceProxy::dataReceived):
* WebProcess/GPU/media/SourceBufferPrivateRemote.cpp:
(WebKit::SourceBufferPrivateRemote::append):
* WebProcess/Network/WebResourceLoader.cpp:
(WebKit::WebResourceLoader::didReceiveData):
* WebProcess/Network/WebResourceLoader.h:
* WebProcess/Network/WebResourceLoader.messages.in:
* WebProcess/Plugins/PDF/PDFPlugin.mm:
(WebKit::PDFPlugin::writeItemsToPasteboard):
* WebProcess/Storage/WebServiceWorkerFetchTaskClient.cpp:
(WebKit::WebServiceWorkerFetchTaskClient::didReceiveData):
(WebKit::WebServiceWorkerFetchTaskClient::didReceiveBlobChunk):
* WebProcess/WebCoreSupport/WebChromeClient.cpp:
* WebProcess/WebCoreSupport/WebEditorClient.cpp:
(WebKit::WebEditorClient::registerAttachmentIdentifier):
* WebProcess/WebCoreSupport/WebPlatformStrategies.cpp:
(WebKit::WebPlatformStrategies::bufferForType):
(WebKit::WebPlatformStrategies::setBufferForType):
(WebKit::WebPlatformStrategies::readBufferFromClipboard):
(WebKit::WebPlatformStrategies::readBufferFromPasteboard):
* WebProcess/WebCoreSupport/mac/WebDragClientMac.mm:
(WebKit::WebDragClient::declareAndWriteDragImage):
* WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::getContentsAsMHTMLData):
(WebKit::WebPage::getSelectionAsWebArchiveData):
(WebKit::WebPage::getMainResourceDataOfFrame):
(WebKit::WebPage::getResourceDataFromFrame):
(WebKit::WebPage::getWebArchiveOfFrame):
(WebKit::WebPage::getAccessibilityTreeData):
(WebKit::WebPage::drawToPDF):
(WebKit::WebPage::drawPagesToPDF):
(WebKit::WebPage::didGetLoadDecisionForIcon):
(WebKit::WebPage::urlSchemeTaskDidReceiveData):
(WebKit::WebPage::updateAttachmentAttributes):
* WebProcess/WebPage/WebPage.h:
* WebProcess/WebPage/WebPage.messages.in:
* WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::getDataSelectionForPasteboard):
(WebKit::WebPage::performActionOnElement):
(WebKit::WebPage::drawToPDFiOS):
(WebKit::WebPage::didFinishLoadForQuickLookDocumentInMainFrame):
* WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::getDataSelectionForPasteboard):
* WebProcess/WebProcess.h:
* WebProcess/WebProcess.messages.in:
* WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::consumeAudioComponentRegistrations):

# Please populate the above commit message. Lines starting
# with &apos;#&apos; will be ignored

# 13  Tim Horton  &lt;timothy_horton@apple.com &gt;
#
# ApplePayLogoSystemImage needlessly loads PassKit in WebKit child processes on iOS
# <a href="https://bugs.webkit.org/show_bug.cgi?id=240404">https://bugs.webkit.org/show_bug.cgi?id=240404</a>
# &lt;rdar://93070349 &gt;
#
# Reviewed by Chris Dumez.
#
# * UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
# (WebKit::ProcessLauncher::launchProcess): Adopt the WTF implementation of systemDirectoryPath.
# (WebKit::systemDirectoryPath): Moved to WTF.
#
# 13  Wenson Hsieh  &lt;wenson_hsieh@apple.com &gt;
#
# ImageAnalysisQueue should reanalyze image elements whose image sources have changed
# <a href="https://bugs.webkit.org/show_bug.cgi?id=240371">https://bugs.webkit.org/show_bug.cgi?id=240371</a>
# rdar://93175651
#
# Reviewed by Tim Horton.
#
# To aid with debugging similar issues in the future, plumb the image URL through to
# `requestImageAnalysisWithIdentifier`, which (if an engineering default is specified) will additionally reveal
# the URL in system logs.
#
# * Platform/cocoa/ImageAnalysisUtilities.h:
# * UIProcess/Cocoa/WebViewImpl.mm:
# (WebKit::WebViewImpl::requestTextRecognition):
# * UIProcess/ios/WKContentViewInteraction.mm:
# (-[WKContentView requestTextRecognition:imageData:identifier:completionHandler:]):
#
# 13  Aditya Keerthi  &lt;akeerthi@apple.com &gt;
#
# [iOS] Multiple visible find highlights when searching for text after beginning a &quot;find from selection&quot;
# <a href="https://bugs.webkit.org/show_bug.cgi?id=240393">https://bugs.webkit.org/show_bug.cgi?id=240393</a>
# rdar://91441895
#
# Reviewed by Wenson Hsieh.
#
# Some WebKit clients use SPI on WKWebView to support &quot;find from selection&quot;
# functionality (the Cmd+E shortcut). However, to support general
# find functionality, they use new find API that uses a different codepath
# to draw highlights. Mixing use of the API and SPI can currently result
# in two highlights showing up.
#
# To fix, ensure SPI highlights are removed once the API is being used.
# The long term solution is for clients to adopt API for the
# &quot;find from selection&quot; functionality, but in the short term the SPI
# should remain supported.
#
# * WebProcess/WebPage/WebPage.cpp:
# (WebKit::WebPage::clearAllDecoratedFoundText):
#
# 13  Kimmo Kinnunen  &lt;kkinnunen@apple.com &gt;
#
# IPC cannot encode, decode Ref&lt; &gt;, causing duplicate code
# <a href="https://bugs.webkit.org/show_bug.cgi?id=240160">https://bugs.webkit.org/show_bug.cgi?id=240160</a>
#
# Reviewed by Antti Koivisto.
#
# Consider reference counted type T:
# Encoding is of signature `encode(const T&amp;)`
# Decoding is of signature `std::optional&lt;Ref&lt;T &gt; &gt; decode()`
# This is usable for both `RefPtr` and `Ref` use-cases.
#
# Add `Ref&lt;T &gt;` encoder that is polymorphic to `T`.
# This way clients do not need to write encoders for
# `Ref&lt;T &gt;`, `RefPtr&lt;T &gt;`. They just need one encoder, for `T`.
#
# When clients do not need to write `RefPtr&lt;T &gt;`, they make less
# bugs. These will be corrected in future patches.
#
# Having `Ref&lt;T &gt;` encoders allows clients to use `Vector&lt;Ref&lt;T &gt; &gt;` encoding
# via `Vector&lt;U &gt;` encoder, and they would not need to duplicate that
# logic, making less bugs. These will be corrected in future patches.
#
# Remove most `IPC::ArgumentCoder&lt;Ref&lt;.. &gt; &gt;` in favor of `IPC::ArgumentCoder&lt;.. &gt;`.
# In the cases where removal was done, remove also redundant
# `IPC::ArgumentCoder&lt;RefPtr&lt;.. &gt; &gt;`.
#
# * Platform/IPC/ArgumentCoders.h:
# (IPC::ArgumentCoder&lt;RefPtr&lt;T &gt; &gt;::decode):
# (IPC::ArgumentCoder&lt;Ref&lt;T &gt; &gt;::encode):
# (IPC::ArgumentCoder&lt;Ref&lt;T &gt; &gt;::decode):
# * Shared/Cocoa/WebCoreArgumentCodersCocoa.mm:
# (IPC::ArgumentCoder&lt;WebCore::Font &gt;::encodePlatformData):
# (IPC::ArgumentCoder&lt;WebCore::Font &gt;::decodePlatformData):
# * Shared/WebCoreArgumentCoders.cpp:
# (IPC::ArgumentCoder&lt;WebCore::Font &gt;::encode):
# (IPC::ArgumentCoder&lt;Font &gt;::decode):
# (IPC::ArgumentCoder&lt;SystemImage &gt;::encode):
# (IPC::ArgumentCoder&lt;SystemImage &gt;::decode):
# * Shared/WebCoreArgumentCoders.h:
# * Shared/playstation/WebCoreArgumentCodersPlayStation.cpp:
# (IPC::ArgumentCoder&lt;Font &gt;::encodePlatformData):
# (IPC::ArgumentCoder&lt;Font &gt;::decodePlatformData):
# * Shared/soup/WebCoreArgumentCodersSoup.cpp:
# (IPC::ArgumentCoder&lt;Font &gt;::encodePlatformData):
# (IPC::ArgumentCoder&lt;Font &gt;::decodePlatformData):
# * Shared/win/WebCoreArgumentCodersWin.cpp:
# (IPC::ArgumentCoder&lt;Font &gt;::encodePlatformData):
# (IPC::ArgumentCoder&lt;Font &gt;::decodePlatformData):
#
# 13  Tim Horton  &lt;timothy_horton@apple.com &gt;
#
# Add UI-side layers for optionally indicating interaction regions
# <a href="https://bugs.webkit.org/show_bug.cgi?id=240372">https://bugs.webkit.org/show_bug.cgi?id=240372</a>
# &lt;rdar://problem/87170289 &gt;
#
# Reviewed by Dean Jackson.
#
# * Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
# (WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToLayer):
# (WebKit::RemoteLayerTreePropertyApplier::applyHierarchyUpdates):
# Allow RemoteLayerTreeInteractionRegionLayers to add and maintain layers for interaction regions.
#
# * SourcesCocoa.txt:
# * UIProcess/RemoteLayerTree/RemoteLayerTreeInteractionRegionLayers.h: Added.
# * UIProcess/RemoteLayerTree/RemoteLayerTreeInteractionRegionLayers.mm: Added.
# (configureLayerForInteractionRegion):
# (WebKit::interactionRegionForLayer):
# (WebKit::isInteractionRegionLayer):
# (WebKit::setInteractionRegion):
# Box WebCore::InteractionRegion in a Objective-C object and store it on the layer.
# Use this as a key to indicate that a given layer is an interaction region.
#
# (WebKit::appendInteractionRegionLayersForLayer):
# Make sure that interaction region layers are always at the end of the layer&apos;s sublayers array.
#
# (WebKit::updateLayersForInteractionRegions):
# Add new layers for interaction regions.
# Maintain the same layer for regions that cover the same area.
# Add a green wash if the default `WKInteractionRegionDebugFill` is set.
#
# * WebKit.xcodeproj/project.pbxproj:
#
# 13  Simon Fraser  &lt;simon.fraser@apple.com &gt;
#
# Simplify the usage of DrawGlyphsRecorder
# <a href="https://bugs.webkit.org/show_bug.cgi?id=240363">https://bugs.webkit.org/show_bug.cgi?id=240363</a>
#
# Reviewed by Myles Maxfield.
#
# Move the DeconstructDrawGlyphs switch from inside DrawGlyphsRecorder to
# DisplayList::Recorder, so that DrawGlyphsRecorder is only ever used when deconstructing
# glyphs.
#
# Also, ensure that the only time that glyph deconstruction is enabled is when
# recording display lists via RemoteDisplayListRecorderProxy. There&apos;s no reason for
# paintAttachmentText() to need glyph deconstruction.
#
# This is a small perf improvement on the MotionMark Design subtest.
#
# * WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.cpp:
# (WebKit::RemoteDisplayListRecorderProxy::RemoteDisplayListRecorderProxy):
# (WebKit::m_renderingBackend):
# * WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.h:
#
# it takes from the XPC message is sent until the XPC reply is received. Add reporting when this happens

# On branch bz/240223
# Your branch is ahead of &apos;origin/main&apos; by 1 commit.
#   (use &quot;git push&quot; to publish your local commits)
#
# Changes to be committed:
#   (use &quot;git restore --staged &lt;file &gt;...&quot; to unstage)
#       modified:   Source/WebKit/ChangeLog
#
# Untracked files:
#   (use &quot;git add &lt;file &gt;...&quot; to include in what will be committed)
#       0001-SharedBuffer-should-be-copied-and-then-attributed-ba.patch
#

Canonical link: <a href="https://commits.webkit.org/250556@main">https://commits.webkit.org/250556@main</a>
git-svn-id: <a href="https://svn.webkit.org/repository/webkit/trunk@294190">https://svn.webkit.org/repository/webkit/trunk@294190</a> 268f45cc-cd09-0410-ab3c-d52691b4dbfc
</pre>

### CODE REVIEW

**Comment by jyavenard** [comment]
Commits will be squashed prior merge. I split them into logical entities to simplify review.

File: Source/WebKit/ChangeLog [hunk 1]
```diff
@@ -466,6 +654,36 @@
         (WebKit::WebPage::handlePDFServiceClick):
         Plumbing for handling a click on the image controls button for PDFs.
 
+2022-05-06  Megan Gardner  <megan_gardner@apple.com>
```
Comment (jernoble) [inline Source/WebKit/ChangeLog]: Somehow, your diff moved Megan's ChangeLog too?

**Review by jernoble — APPROVED** [review]
LGTM. Most changes are a SharedBufferCopy -> SharedBufferReference rename. Meat of the change appears to be within the new SharedBufferReference class. r=me

File: Source/WebKit/ChangeLog (hunk 1 above)
Comment (jyavenard) [reply Source/WebKit/ChangeLog]: Changelog rebase is completely broken. That’s not my doing.

**Pull request merged**